File: src/node/vaonis_instruments_unified/src/unified_client.js
import { readFileSync } from "fs";
import { resolve, dirname } from "path";
import { fileURLToPath } from "url";

const __dirname = dirname(fileURLToPath(import.meta.url));
const ROUTES_PATH = resolve(__dirname, "../data/http_routes_union.json");

export const DEFAULT_PREFIXES = ["", "/stellina/http", "/vespera/http"];

let routesCache = null;

const ANSI_RESET = "\x1b[0m";
const ANSI_BLUE = "\x1b[94m";
const ANSI_GREEN = "\x1b[92m";
const ANSI_CYAN = "\x1b[96m";
const ANSI_MAGENTA = "\x1b[95m";

/**
 * @typedef {Object} RouteDefinition
 * @property {string} operationId
 * @property {string} method
 * @property {string} path
 */

/**
 * @typedef {Object} RequestOptions
 * @property {string|null} [auth]
 * @property {Record<string, any>|null} [params]
 * @property {any} [jsonBody]
 * @property {any} [body]
 * @property {Record<string, string>} [headers]
 */

/**
 * @typedef {Object} DownloadImageOptions
 * @property {string|null} [auth]
 * @property {Record<string, string>} [headers]
 * @property {Record<string, any>|null} [params]
 */

/**
 * @typedef {Object} ClientOptions
 * @property {string} [host]
 * @property {number} [port]
 * @property {string} [apiBasePath]
 * @property {Iterable<string>} [prefixes]
 * @property {number} [timeoutMs]
 * @property {typeof fetch} [fetchImpl]
 * @property {{ log: (...args: any[]) => void }} [logger]
 * @property {boolean} [logResponses]
 * @property {boolean} [colorizeResponses]
 * @property {number} [logResponseLimit]
 */

/**
 * @returns {Map<string, RouteDefinition>}
 */
function loadRoutes() {
  if (routesCache) {
    return routesCache;
  }
  const data = JSON.parse(readFileSync(ROUTES_PATH, "utf-8"));
  routesCache = new Map(data.map((item) => [item.operationId, item]));
  return routesCache;
}

/**
 * @param {URL} url
 * @param {Record<string, any>|null|undefined} params
 * @returns {URL}
 */
function buildQuery(url, params) {
  if (!params) {
    return url;
  }
  for (const [key, value] of Object.entries(params)) {
    if (Array.isArray(value)) {
      for (const item of value) {
        url.searchParams.append(key, String(item));
      }
    } else if (value !== undefined && value !== null) {
      url.searchParams.set(key, String(value));
    }
  }
  return url;
}

/**
 * @param {Buffer} buffer
 * @param {string} contentType
 * @returns {boolean}
 */
function looksLikeImage(buffer, contentType) {
  if (contentType && contentType.toLowerCase().startsWith("image/")) {
    return true;
  }
  if (!buffer || buffer.length < 4) {
    return false;
  }
  if (buffer[0] === 0xff && buffer[1] === 0xd8 && buffer[2] === 0xff) {
    return true; // JPEG
  }
  if (buffer.slice(0, 8).equals(Buffer.from([0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a]))) {
    return true; // PNG
  }
  if (buffer.slice(0, 2).equals(Buffer.from("BM"))) {
    return true; // BMP
  }
  if (buffer.slice(0, 6).equals(Buffer.from("GIF87a")) || buffer.slice(0, 6).equals(Buffer.from("GIF89a"))) {
    return true; // GIF
  }
  if (buffer.slice(0, 4).equals(Buffer.from([0x49, 0x49, 0x2a, 0x00])) || buffer.slice(0, 4).equals(Buffer.from([0x4d, 0x4d, 0x00, 0x2a]))) {
    return true; // TIFF
  }
  if (buffer.length >= 12 && buffer.slice(0, 4).equals(Buffer.from("RIFF")) && buffer.slice(8, 12).equals(Buffer.from("WEBP"))) {
    return true; // WEBP
  }
  return false;
}

function colorizeJsonLine(line) {
  const match = line.match(/^(\s*)("[^"]+")(\s*:\s*)(.*)$/);
  if (!match) {
    return line;
  }
  const [, indent, key, sep, rest] = match;
  const keyColored = `${ANSI_BLUE}${key}${ANSI_RESET}`;
  let value = rest;
  const stringMatch = value.match(/^"[^"]*"/);
  if (stringMatch) {
    const token = stringMatch[0];
    value = `${ANSI_GREEN}${token}${ANSI_RESET}${value.slice(token.length)}`;
  } else {
    const numberMatch = value.match(/^-?\d+(?:\.\d+)?(?:[eE][+-]?\d+)?/);
    if (numberMatch) {
      const token = numberMatch[0];
      value = `${ANSI_CYAN}${token}${ANSI_RESET}${value.slice(token.length)}`;
    } else {
      const boolMatch = value.match(/^(true|false|null)\b/);
      if (boolMatch) {
        const token = boolMatch[0];
        value = `${ANSI_MAGENTA}${token}${ANSI_RESET}${value.slice(token.length)}`;
      }
    }
  }
  return `${indent}${keyColored}${sep}${value}`;
}

function colorizeJson(text) {
  return text.split("\n").map(colorizeJsonLine).join("\n");
}

function formatPayload(value, { color = false, maxLen = 4000 } = {}) {
  let rendered = "";
  if (Buffer.isBuffer(value)) {
    rendered = `<${value.length} bytes>`;
  } else if (value && typeof value === "object") {
    rendered = JSON.stringify(value, null, 2);
  } else if (typeof value === "string") {
    const trimmed = value.trim();
    if (trimmed) {
      try {
        const parsed = JSON.parse(trimmed);
        rendered = JSON.stringify(parsed, null, 2);
      } catch {
        rendered = value;
      }
    } else {
      rendered = value;
    }
  } else {
    rendered = String(value);
  }
  if (maxLen > 0 && rendered.length > maxLen) {
    rendered = `${rendered.slice(0, maxLen - 20)}... (truncated)`;
  }
  if (color) {
    return colorizeJson(rendered);
  }
  return rendered;
}

export class UnifiedHttpClient {
  /**
   * @param {ClientOptions} [options]
   */
  constructor({
    host = "10.0.0.1",
    port = 8082,
    apiBasePath = "/v1",
    prefixes = DEFAULT_PREFIXES,
    timeoutMs = 10000,
    fetchImpl = globalThis.fetch,
    logger = console,
    logResponses = true,
    colorizeResponses = true,
    logResponseLimit = 4000,
  } = {}) {
    if (!fetchImpl) {
      throw new Error("fetch is required (Node 18+ or provide fetchImpl)");
    }
    this.host = host;
    this.port = port;
    this.apiBasePath = apiBasePath;
    this.prefixes = Array.from(prefixes);
    this.timeoutMs = timeoutMs;
    this.fetch = fetchImpl;
    this.baseUrl = null;
    this.logger = logger;
    this.logResponses = logResponses;
    this.colorizeResponses = colorizeResponses;
    this.logResponseLimit = logResponseLimit;
  }

  /**
   * @returns {Promise<string>}
   */
  async detectBaseUrl() {
    for (const prefix of this.prefixes) {
      const candidate = this._makeBaseUrl(prefix);
      try {
        const response = await this._fetchWithTimeout(`${candidate}/app/status`, {
          method: "GET",
        });
        if (response.ok) {
          this.baseUrl = candidate;
          return candidate;
        }
      } catch {
        // Ignore and try next prefix.
      }
    }
    throw new Error("Unable to detect a working base URL");
  }

  /**
   * @param {string} method
   * @param {string} path
   * @param {RequestOptions} [options]
   * @returns {Promise<any>}
   */
  async request(
    method,
    path,
    { auth = null, params = null, jsonBody = undefined, body = undefined, headers = {} } = {}
  ) {
    const base = this.baseUrl || (await this.detectBaseUrl());
    const url = new URL(`${base.replace(/\/+$/, "")}/${path.replace(/^\//, "")}`);
    buildQuery(url, params);

    const resolvedHeaders = { ...headers };
    if (auth) {
      resolvedHeaders.Authorization = auth;
    }
    let resolvedBody = body;
    if (jsonBody !== undefined) {
      resolvedBody = JSON.stringify(jsonBody);
      if (!resolvedHeaders["Content-Type"]) {
        resolvedHeaders["Content-Type"] = "application/json";
      }
    }

    const response = await this._fetchWithTimeout(url.toString(), {
      method,
      headers: resolvedHeaders,
      body: resolvedBody,
    });

    const contentType = response.headers.get("content-type") || "";
    let payload;
    if (contentType.includes("application/json")) {
      payload = await response.json();
    } else {
      payload = await response.text();
    }

    if (this.logResponses) {
      const formatted = formatPayload(payload, {
        color: this.colorizeResponses,
        maxLen: this.logResponseLimit,
      });
      this.logger.log(`HTTP ${response.status} ${response.statusText}\n${formatted}`);
    }

    if (!response.ok) {
      const message = typeof payload === "string" ? payload : "";
      throw new Error(`HTTP ${response.status} ${response.statusText} ${message}`.trim());
    }
    return payload;
  }

  /**
   * @param {string} operationId
   * @param {RequestOptions} [options]
   * @returns {Promise<any>}
   */
  async callOperation(operationId, options = {}) {
    const route = loadRoutes().get(operationId);
    if (!route) {
      throw new Error(`Unknown operationId: ${operationId}`);
    }
    return this.request(route.method, route.path, options);
  }

  /**
   * @param {string} urlOrPath
   * @param {DownloadImageOptions} [options]
   * @returns {Promise<Buffer>}
   */
  async downloadImage(urlOrPath, { auth = null, headers = {}, params = null } = {}) {
    const base = this.baseUrl || (await this.detectBaseUrl());
    const fullUrl =
      urlOrPath.startsWith("http://") || urlOrPath.startsWith("https://")
        ? urlOrPath
        : `${base.replace(/\/+$/, "")}/${urlOrPath.replace(/^\//, "")}`;
    const url = new URL(fullUrl);
    buildQuery(url, params);

    const resolvedHeaders = { ...headers };
    if (auth) {
      resolvedHeaders.Authorization = auth;
    }
    const response = await this._fetchWithTimeout(url.toString(), {
      method: "GET",
      headers: resolvedHeaders,
    });
    const buffer = Buffer.from(await response.arrayBuffer());
    const contentType = response.headers.get("content-type") || "";
    if (this.logResponses) {
      const formatted = formatPayload(buffer, {
        color: false,
        maxLen: this.logResponseLimit,
      });
      this.logger.log(
        `HTTP ${response.status} ${response.statusText} image bytes=${formatted}`
      );
    }
    if (response.ok || looksLikeImage(buffer, contentType)) {
      return buffer;
    }
    throw new Error(`HTTP ${response.status} ${response.statusText}`.trim());
  }

  /**
   * @param {string} prefix
   * @returns {string}
   */
  _makeBaseUrl(prefix) {
    const raw = `http://${this.host}:${this.port}${prefix}${this.apiBasePath}`;
    return raw.replace(/\/+$/, "");
  }

  /**
   * @param {string} url
   * @param {RequestInit} options
   * @returns {Promise<Response>}
   */
  async _fetchWithTimeout(url, options) {
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), this.timeoutMs);
    try {
      return await this.fetch(url, { ...options, signal: controller.signal });
    } finally {
      clearTimeout(timeout);
    }
  }
}

export { loadRoutes };

File: src/python/vaonis_instruments/unified_client.py
"""Unified HTTP client covering firmware routes.

This client is intended to call any route listed in the firmware union manifest.
It can autodetect base path prefixes and call routes by operationId.
"""

from __future__ import annotations

from dataclasses import dataclass, field
import json
import logging
import time
from pathlib import Path
from typing import Any, Dict, Iterable, Optional

import requests

from .logging_utils import format_payload


def _looks_like_image(data: bytes, content_type: Optional[str]) -> bool:
    if content_type and content_type.lower().startswith("image/"):
        return True
    if len(data) < 4:
        return False
    if data[:3] == b"\xff\xd8\xff":
        return True  # JPEG
    if data[:8] == b"\x89PNG\r\n\x1a\n":
        return True  # PNG
    if data[:2] == b"BM":
        return True  # BMP
    if data[:6] in (b"GIF87a", b"GIF89a"):
        return True  # GIF
    if data[:4] in (b"II*\x00", b"MM\x00*"):
        return True  # TIFF
    if len(data) >= 12 and data[:4] == b"RIFF" and data[8:12] == b"WEBP":
        return True  # WEBP
    return False


DEFAULT_PREFIXES = ["", "/stellina/http", "/vespera/http"]


@dataclass
class UnifiedHTTPClient:
    host: str = "10.0.0.1"
    port: int = 8082
    api_base_path: str = "/v1"
    prefixes: Iterable[str] = field(default_factory=lambda: list(DEFAULT_PREFIXES))
    timeout_s: float = 10.0
    session: requests.Session = field(default_factory=requests.Session)
    base_url: Optional[str] = None
    logger: Optional[logging.Logger] = None
    log_payloads: bool = True
    log_payload_limit: int = 4000

    def _log(self, message: str, *args: Any) -> None:
        if self.logger is not None:
            self.logger.info(message, *args)

    def detect_base_url(self) -> str:
        for prefix in self.prefixes:
            candidate = self._make_base_url(prefix)
            try:
                resp = self.session.get(
                    f"{candidate}/app/status", timeout=self.timeout_s
                )
                if resp.ok:
                    self.base_url = candidate
                    return candidate
            except requests.RequestException:
                continue
        raise RuntimeError("Unable to detect a working base URL")

    def request(
        self,
        method: str,
        path: str,
        *,
        auth: Optional[str] = None,
        params: Optional[Dict[str, Any]] = None,
        json_body: Any = None,
        files: Any = None,
        stream: bool = False,
    ) -> Any:
        base = self.base_url or self.detect_base_url()
        url = base.rstrip("/") + "/" + path.lstrip("/")
        headers: Dict[str, str] = {}
        if auth is not None:
            headers["Authorization"] = auth
        body_repr: Optional[str] = None
        if json_body is not None:
            try:
                body_repr = json.dumps(json_body, ensure_ascii=True)
            except TypeError:
                body_repr = str(json_body)
        self._log(
            "HTTP request %s %s params=%s json=%s",
            method,
            url,
            params,
            body_repr,
        )
        start = time.monotonic()
        try:
            resp = self.session.request(
                method,
                url,
                params=params,
                json=json_body,
                files=files,
                headers=headers,
                timeout=self.timeout_s,
                stream=stream,
            )
        except requests.RequestException as exc:
            if self.logger is not None:
                self.logger.exception("HTTP error %s %s: %s", method, url, exc)
            raise
        elapsed_ms = (time.monotonic() - start) * 1000.0
        size: Optional[int]
        if stream:
            size = None
        else:
            try:
                size = len(resp.content)
            except Exception:
                size = None
        self._log(
            "HTTP response %s %s status=%s elapsed_ms=%.1f bytes=%s",
            method,
            url,
            resp.status_code,
            elapsed_ms,
            size,
        )
        if self.logger is not None and self.log_payloads and not stream:
            content_type = resp.headers.get("content-type", "")
            if (
                content_type.startswith("image/")
                or "application/octet-stream" in content_type
            ):
                payload = f"<{len(resp.content)} bytes>"
            else:
                payload = format_payload(
                    resp.text, color=False, max_len=self.log_payload_limit
                )
            self._log("HTTP response body %s %s payload=%s", method, url, payload)
        resp.raise_for_status()
        try:
            return resp.json()
        except ValueError:
            return resp.text

    def call_operation(
        self,
        operation_id: str,
        *,
        auth: Optional[str] = None,
        params: Optional[Dict[str, Any]] = None,
        json_body: Any = None,
        files: Any = None,
        stream: bool = False,
    ) -> Any:
        route = _load_routes().get(operation_id)
        if route is None:
            raise KeyError(f"Unknown operationId: {operation_id}")
        return self.request(
            route["method"],
            route["path"],
            auth=auth,
            params=params,
            json_body=json_body,
            files=files,
            stream=stream,
        )

    def download_image(
        self, url: str, stream: bool = False, auth: Optional[str] = None
    ) -> Any:
        base = self.base_url or self.detect_base_url()
        if url.startswith("http://") or url.startswith("https://"):
            full_url = url
        else:
            full_url = base.rstrip("/") + "/" + url.lstrip("/")
        self._log("HTTP image request GET %s", full_url)
        headers: Dict[str, str] = {}
        if auth is not None:
            headers["Authorization"] = auth
        resp = self.session.get(
            full_url, timeout=self.timeout_s, stream=stream, headers=headers
        )
        if stream:
            return resp
        content = resp.content
        content_type = resp.headers.get("content-type")
        self._log(
            "HTTP image response %s status=%s bytes=%s",
            full_url,
            resp.status_code,
            len(content),
        )
        if resp.ok or _looks_like_image(content, content_type):
            return content
        resp.raise_for_status()
        return content

    def _make_base_url(self, prefix: str) -> str:
        return f"http://{self.host}:{self.port}{prefix}{self.api_base_path}".rstrip("/")


_ROUTES_CACHE: Optional[Dict[str, Dict[str, Any]]] = None


def _load_routes() -> Dict[str, Dict[str, Any]]:
    global _ROUTES_CACHE
    if _ROUTES_CACHE is not None:
        return _ROUTES_CACHE
    routes_path = Path(__file__).resolve().parent / "data" / "http_routes_union.json"
    data = json.loads(routes_path.read_text(encoding="utf-8"))
    _ROUTES_CACHE = {item["operationId"]: item for item in data}
    return _ROUTES_CACHE


__all__ = ["UnifiedHTTPClient"]
